Resolve #1: HttpUrlConnectionExecutor appends the system http agent t…

lemonboston opened the pull request:
…o the User-Agent header.

lemonboston commented:
Should it be optimized further to do the append (or save the current user-agent header value) during
`for (Header<?> header : request.headers())`,
to avoid iteration through headers again in
`connection.getRequestProperty("User-Agent")`?
(underlying values stored in arrays, not in map)

Code readability vs performance question here, I am not sure how much we need to optimize.

dmfs commented:
It's not a performance issue (the actual network request is probably an order of magnitude slower).

Why not just do that inside of the loop?
like so:
```
if ("user-agent".equalsignorecase(header.type().name()))
{
    // append header
}
else
{
    // don't append header
}
```
That won't append the user-agent if there is none in the headers, but I actually would consider that ok. We should not enforce a user-agent header but only extend it if there is any.

lemonboston commented:
Yes, appending inside the loop was a possible variation and if we don't need to add the system agent, it's also simpler.
Before I update I just want to ask why is better to not add the lone system http agent? Isn't there a chance that it became useful for the server or just for statistics? I know it's not browser and webpage, where it is important, but who knows if we needed something to tweek on the server based on the Android version or phone brand. (I admit that it seems very unlikely though :)) But in general, why not send a standard information if we can?
(This was returned for my phone: `Dalvik/2.1.0 (Linux; U; Android 6.0.1; Nexus 5 Build/MOB30Y)`
I am okay with any way of course, so if you confirm I'll update it.

dmfs commented:
I think we should not enforce a user-agent if a client prefers to operate in "stealth mode" (i.e. doesn't any user-agent on its own). We have no way to suppress the generation of this header otherwise.

Good point about the name of the Executor implementation. Ideally we have BuildConfig to provide these values. I've created a ticket #10.

lemonboston commented:
Updated to only append system agent when there is already a user-agent header.
@dmfs please check

dmfs commented:
Also consider to wrap `BuildConfic.NAME` and `BuildConfig.VERSION` in `SafeToken`s before adding it to the user-agent. Just to be sure the result is always valid.

lemonboston commented:
Putting `SYSTEM_HTTP_AGENT != null` second was deliberate, because I expect that it will almost never be null. I think I read somewhere System.getProperty("http.agent") is available on something like Android 2.1+. So for these devices putting as second seems optimal to avoid checking it unnecessarily for each header.

I am going to rebase with master and add the name and version using BulidConfig.

lemonboston commented:
Updated, so both build name, version and system http agent gets appended now.
(note: rebased onto master and force pushed back to origin)
Please review.

dmfs commented:
Hmm... I'm not a big friend of static code blocks. Heavy work should not be done in a static context. Anyway, we'll keep it like that for now. Maybe we refactor this later on.

Please add a test case. HttpUrlConnection is abstract, so you can just write a factory that returns a MockHttpUrlConnection.

lemonboston commented:
You are right, after a second look I agree that it should not be in static block. I moved it to constructor.
I've also added the functional test for HttpUrlConnectionExecutor that checks the User-Agent header appending.
Please review.

dmfs commented on httpurlconnection-executor/src/test/java/org/dmfs/httpessentials/httpurlconnection/EmptyRequestWithHeaders.java:
```diff
@@ -0,0 +1,88 @@
+    @Override
+    public HttpRequestEntity requestEntity()
+    {
+        return new HttpRequestEntity()
```
In http-client-basics we already have a class EmptyRequestEntity, check it out.

lemonboston replied:
Thanks, updated to use it. (Needed to add the module as test dependency.)

dmfs commented on httpurlconnection-executor/src/test/java/org/dmfs/httpessentials/httpurlconnection/RequestPropertyCapturingURLConnection.java:
```diff
@@ -0,0 +1,53 @@
+ *
+ * @author Gabor Keszthelyi
+ */
+public final class RequestPropertyCapturingURLConnection extends HttpURLConnection
```
I wonder if we can evolve that into a MockHttpUrlConnection that essentially just returns the headers and request entity that has been sent. Just like a server that mirrors the request back as the response. That way we could use this class for other test cases in this library as well.
The ResponseHandler of the request would just filter out the value under test, i.e. it would return the string value of the user-agent header in the response. That means you can just test the value returned by `execute`.
Do you know what I mean?

lemonboston replied:
I kind of understand the analogy with the mirroring server, but for me using the return value of a method to retrieve some captured values from inside the execution would be messing things up. I mean the 'semantics', or not sure how to call it, would be altered. For example if somebody else would look into the test and see assertion on the return value checking against something else, it would not be clear.
Also, it would not be a general approach, for example it couldn't be applied to void methods or methods with String return value for example.

So I think in the area of these 'code' tests (not servers) the concept of capturing is clearer, that is we catch some argument values for later assertion.
If we want to make a more general class of this for reuse, we can add capturing to every method/argument we are interested in (so we have more captured fields). And if we wanted to make assertions simpler for convenience - the filtering part you mentioned (if I understood that correctly) - I would probably just add methods for that to the class, like
```
public void verifyPropertyWasSet(String key, String expectedValue)
{
   // check in capturedProperties and assert
}
```
I assume you may have concerns relating to immutability, encapsulation with these captured fields..  I think it is fine here, not the same consideration applies as to production code, it's basically a tool for testing, a stub for a specific class.

lemonboston replied:
Sorry, I probably didn't read your suggestion carefully first...  I see you would want to test the full circle, so to say, so the HttpUrlConnectionResponse as well...
I will get back to this tomorrow.

dmfs replied:
Yes, the idea is that MockHttpUrlConnection emulates a mirroring server that returns the entire request as is (it returns the same headers and same entity content). Then we use HttpResponseHandler to filter and return the data that is tested.
That allows us to test every aspect of HttpUrlConnectionExecutor without needing an actual server.

I don't want to test the full circle. HttpUrlConnectionResponse is not tested (but involved) in this case and it should be tested separately. Given it's well tested we can assume that it behaves correctly when it's involved in other test cases (like in this case).

lemonboston replied:
So as I understand these 'steps' would be involved in a test execution:

0. setting up the custom ResponseHandler for the data we are interested in
execution:
1. adapting HttpRequest to HttpUrlConnection
2. MockHttpUrlConnection mirrors its inputs to produce the same response
3. adapting HttpUrlConnection to HttpResponse
4. custom ResponseHandler takes configured data from HttpResponse and returns it
verification:
5. asserting execute() return value against expected value

So we would always test 1. and 3. together, and even if as you say we would trust 3. because it's tested separately, I still think this is not good. With any of these tests we probably only want to check either 1. or 3., executing both would make the intent of the test not clear. 
And with 4. and 5. we would add some completely arbitrary heck to make possible to access the value under test as the return value of the public method that normally represents something else, which would again make things unclear if somebody just looks into the test.
So as you said that 3. would be tested in isolation, I think it much clearer and straightforward if 1. is just tested separately as well.

dmfs replied:
No, we would not always test 1. and 3.. These are just tools we use. Tools that (should) have already been well tested. Saying that we would always test 1. and 3. is like saying that you also always test junit, mockito, the JVM or Dalvik when you run a test. Or in this case it's like saying that you're also testing HashMap.

I totally agree that this is not how you usually would write a test, but I don't see any flaws with it. It's just another tool that you need to be familiar with before you understand what's going on. Just like jmockit. When you look at a JMockit test for the very first time, it's pretty hard to tell what's actually going on.

The important part is that it makes writing tests easier, IMO. Mostly because you don't need to subclass HttpUrlConnection for every test. The easier something is, the harder is it to make mistakes.

lemonboston replied:
I have a feeling that we won't convince each other on this :)

I don't see 3. as a tool like junit, mockito, JVM, HashMap. Those are external to our code, 3rd party/framework/system stuff that we always treat as trusted and correct, and they also don't carry 'domain logic' so to say.
3. (=HttpUrlConnectionResponse) is our production code which participates with very similar role and has similar levels of complexity as 1. in this execution.
It's okay to write functional/integration tests that includes multiple units/components, but this wouldn't be a functional test in that meaning, because as you said we would use 3. as a tool, and we would not include it with the intention to test a real scenario end2end. If we wanted to test if data comes back from a mirroring server as expected, then it would be. But here we want to check if we correctly adapt our Http API to HttpUrlConnection API. So we are actually interested in the values and calls HttpUrlConnection receive. By adding steps 2. , 3. we convert those values and assert on far derivatives in the end, not on the actual values we wanted to check.

Also, I don't see how creating such a special purpose mini testing framework with limited scope, with non-standard 'semantics' that should probably be documented, is worth creating. Maybe if you shared what other tests you had in mind, I could understand the motives better.

Regarding JMockit I agree that it is far from trivial at first, and the decision to use it or not is not trial also. But there the decision is about using a general-purpose, powerful external mocking tool which once learnt can be used in any test.

But let me know what you would prefer, I can work on implementing that approach if you want.

lemonboston replied:
> here we want to check if we correctly adapt our Http API to HttpUrlConnection API

I just realized that we don't only do this in HttpUrlConnectionExecutor currently, because we also modify the data now for the user-agent header. So I think user-agent appending might be better as a decorator here as well after all. HttpUrlConnectionExecutor should only really be an adapter between the APIs, should not do anything else probably.
What do you think?

lemonboston replied:
The decoration could be hidden from the client of course, it would come bundled. (Factory, or decorating in constructor)

dmfs commented:
I don't think a decorator pattern will work here. The HttpUrlConnection class is not well designed for this. In particular it's not immutable, carries a (quite complex) state and has a gazillion of methods that you would have to override, if possible after all. If anything, you would have to decorate the Headers. 

Also, I don't see why we should have less trust in our own code that in any 3rd party code we use. If we don't trust our own code as much as we trust the dependencies we're doing something wrong. In particular I would never publish any code that I don't trust.

Anyway, I won't insist on the MockHttpUrlConnection. Please let me know when you think you're done with this and this is ready for review.

lemonboston commented:
Sorry, I missed out that I was referring to decorating the HttpRequestExecutor, not HttpUrlConnection.
So just as with Branded, only with appending at the end.
Current (on master) HttpUrlConnectionExecutor would be responsible purely for adapting dmfs Http API to HttpUrlConnection, without modifying any value.
And user-agent header appending would be done with decoration. The same approach as you envisioned with all these decorators, that separate aspects can be applied separately, no need to put everything into one class.
Benefit would be that header append would be done with our API, so it can be re-used if other http client lib adapter executors are created later.

Client would not see this, decoration could be done in constructor, something like this:
```
    public HttpUrlConnectionExecutor(final HttpUrlConnectionFactory connectionFactory)
    {
        Product httpClientProduct = new VersionedProduct(new SafeToken(BuildConfig.NAME), new SafeToken(BuildConfig.VERSION));
        Product systemHttpAgentProduct = new SafeProduct(System.getProperty("http.agent"));
        mExecutor = new Branded(new Branded(new HttpUrlConnectionAdapterExecutor(connectionFactory), systemHttpAgentProduct, true), httpClientProduct, true);
    }

    @Override
    public <T> T execute(final URI uri, final HttpRequest<T> request)            
    {
        return mExecutor.execute(uri, request);        
    }
```
(HttpUrlConnectionAdapterExecutor is what HttpUrlConnectionExecutor is currently (on master))
Or we could have a static factory method, like 
`static HttpRequestExecutor HttpUrlConnectionExecutorFactory.create()`.
As I see it, in general, static methods should be avoided in favor of instance ones but with this sort of entry point to a library, they can actually help decoupling, because getting an instance won't be tied to a certain constructor of a certain subclass, future releases can do whatever they want under this factory method. This means we kind of draw a boundary with this for this module, this is the point where clients should get their executor instance that uses HttpUrlConnection. This can allow the flexibility to add anything later, if needed (like other decorators).
(It's actually exactly what Factory pattern's intent is about as in [here](http://www.oodesign.com/factory-pattern.html).)

For Branded, the extra boolean param to indicate appending in the end could be added, while keeping the other constructor that defaults to append in the beginning, and documenting it well. Or if we want to hide this usage from clients (although, I am not sure why is that needed, maybe they would need this to implement, modify their executor adapters of their own), then we could make a separate Branded with different name (there are just a few lines in it anyway).

We would need SafeProduct in this case for the system agent string, it would replace/escape characters as SafeComment does. Or RawStringProduct which would not do anything. (Currently we don't make this string safe either.)

Separating API adaption and agent-header append this way would make the testing easier as well.

What do you think?

(I think this would be a better way, but current version on this branch is working as well, so if you want to just make progress, we could merge this after review, and create separate enhancement issue. But I am happy to implement it next if you agree with the change.)

dmfs commented:
Good one. You're completely right. Just a few thoughts:

* We'll go with two executors:
 1. An unbranded one that doesn't add its own user agent (essentially the one we already have) 
 2. A branded one that decorates the unbranded one and just delegates to that
* Both executors will be public, so users can choose (again, there might be a reason to have your client operating in "stealth mode", may it be for privacy reasons).
* Instead of using `SafeProduct` or something like `RawStringProduct` for the platform user agent, we would have a `Platform` which implements `Product` and takes no parameters (because it reads `http.agent` internally).
So it would look like
```java
Product systemHttpAgentProduct = new Platform();
```
I think we don't have to validate the platform user-agent. We trust that the platform returns a valid value.
* We're not going to add that boolean parameter to `Branded`. The new decorator should be "internal" to `HttpUrlConnectionExecutor`. I want to make sure that users do not add their user-agent to the end. Only Executors should to that. Maybe we should create a module `http-client-executor-utils` with tools that only Executors need.

dmfs commented:
Since the new approach is fairly different, I suggest to branch from master and start a new PR.
